otel-agent/ecs-ec2/ecsattributesprocessor/docker.go

```diff
@@ -0,0 +1,186 @@
+package ecsattributesprocessor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/coralogix/telemetry-shippers/otel-agent/ecs-ec2/ecsattributesprocessor/internal/metadata"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/client"
+	"go.uber.org/zap"
+)
+
+const (
+	ecsContainerMetadataURIv4 = "ECS_CONTAINER_METADATA_URI_V4"
+	ecsContainerMetadataURI   = "ECS_CONTAINER_METADATA_URI"
+)
+
+type metadataHandler struct {
+	sync.Mutex
+	// endpoints map[string][]string
+	metadata map[string]Metadata
+
+	// logger
+	logger *zap.Logger
+
+	// endpoints -  function used to fetch metadata endpoints
+	endpoints endpointsFn
+
+	stop chan struct{}
+}
+
+func (m *metadataHandler) get(key string) (Metadata, bool) {
+	// initial check for metadata
+	val, ok := m.metadata[key]
+	return val, ok
+}
+
+func (m *metadataHandler) syncMetadata(ctx context.Context, endpoints map[string][]string) error {
+
+	for k, v := range endpoints {
+		if _, ok := m.metadata[k]; ok {
+			continue // if metadata already exists, skip
+		}
+
+		var metadata Metadata
+		resp, err := http.Get(v[0]) // use the 1st available endpoint
+		if err != nil {
+			return fmt.Errorf("failed while calling metadata endpoint for [%s] - %w", k, err)
+		}
+
+		if err = json.NewDecoder(resp.Body).Decode(&metadata); err != nil {
+			return fmt.Errorf("failed to decode metadata for [%s] - %w", k, err)
+		}
+
+		m.metadata[k] = metadata
+	}
+
+	// remove keys that don't exist in current endpoint view
+	// Note: this insures that the amount of metadata we store stays
+	// in line with the number of containers discovered and prevents
+	// us accumulating unused metadata indifinitely
+	for k := range m.metadata {
+		if _, ok := endpoints[k]; !ok {
+			delete(m.metadata, k)
+		}
+	}
+
+	return nil
+}
+
+func (m *metadataHandler) start() error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		m.logger.Sugar().Errorf("failed to intial docker API client: %ss", err)
+		return err
+	}
+
+	go func() {
+		ctx := context.Background()
+		ticker := time.NewTicker(time.Second * 60)
+
+		dockerEvents, errors := cli.Events(context.Background(), types.EventsOptions{})
+
+		for {
+			select {
+			case <-ticker.C:
+				syncMetadata(ctx, m)
+
+			case event := <-dockerEvents:
+				if event.Type == events.ContainerEventType && event.Action == "create" {
+					m.logger.Debug("new container id detected, re-syncing metadata", zap.String("id", event.ID))
+					syncMetadata(ctx, m)
+				}
+
+			case err := <-errors:
+				m.logger.Sugar().Errorf("error received from docker container events: %w", err)
+
+			case <-m.stop:
+				m.logger.Debug("stopping metadata sync")
+				return
+			}
+		}
```

Comment:
Sure, will update the doc. 

I didn't  return the error upstream here because that would crash otel. Since this action is being run every 60s, there is a possibility that there could be errors, but also that these errors would self correct on the next iteration.